File: lib/data/source/sensors/garmin/garmin_file_store.dart
import 'dart:io';

import 'garmin_log.dart';
import 'garmin_session.dart';

/// Keeps the raw FIT files pulled off a watch.
///
/// **This exists because archiving is destructive from our side.** Telling the
/// watch a file is archived makes it stop offering that file forever; if the
/// importer then mis-reads it — as it demonstrably can — the data cannot be
/// fetched again to re-import, because it is gone from the watch and was never
/// kept. Exactly that happened to a night of sleep during development.
///
/// So the contract is: bytes land on disk BEFORE the archive flag is sent, and a
/// failure to write means the file is not archived and the watch offers it again
/// next time. Gadgetbridge keeps the same guarantee by exporting every download.
///
/// A side benefit is diagnosis: a parser bug can be reproduced against the exact
/// bytes that caused it, with no watch in the loop.
class GarminFileStore {
  const GarminFileStore({
    required this.resolveDirectory,
    this.retention = const Duration(days: 30),
  });

  /// Resolves where files are written, on first use.
  ///
  /// A callback rather than a [Directory] because the app's documents directory
  /// is only available asynchronously, and making the whole provider async to
  /// get it would push a `FutureProvider` through every caller. Injected so a
  /// test can point it at a temp dir.
  final Future<Directory> Function() resolveDirectory;

  /// How long a file is kept before [prune] removes it. FIT files are small
  /// (a night of sleep is under a kilobyte, a day of monitoring a few) but the
  /// watch produces them daily and forever is not a retention policy.
  final Duration retention;

  /// Writes [file] and returns its path.
  ///
  /// Throws on failure — deliberately, because the caller's whole reason to
  /// await this is to decide whether archiving is safe.
  Future<String> save(GarminDownloadedFile file, {required DateTime now}) async {
    final directory = await resolveDirectory();
    await directory.create(recursive: true);
    // Type and index identify it; the timestamp keeps a re-download from
    // clobbering an earlier copy, since several files share file number 65535.
    final name = '${file.entry.type.name}_${file.entry.fileIndex}_'
        '${now.toUtc().millisecondsSinceEpoch}.fit';
    final path = '${directory.path}${Platform.pathSeparator}$name';
    await File(path).writeAsBytes(file.bytes, flush: true);
    garminLog('[GARMIN-STORE] saved $name (${file.bytes.length}B)');
    return path;
  }

  /// Deletes files older than [retention]. Best-effort: housekeeping must never
  /// fail a sync.
  Future<void> prune({required DateTime now}) async {
    final cutoff = now.subtract(retention);
    try {
      final directory = await resolveDirectory();
      if (!directory.existsSync()) return;
      await for (final entity in directory.list()) {
        if (entity is! File || !entity.path.endsWith('.fit')) continue;
        final stat = await entity.stat();
        if (stat.modified.isBefore(cutoff)) {
          await entity.delete();
          garminLog('[GARMIN-STORE] pruned ${entity.path}');
        }
      }
    } catch (error) {
      garminLog('[GARMIN-STORE] prune failed: $error');
    }
  }
}

File: app/src/main/kotlin/tech/mmarca/openvitals/devices/garmin/GarminFileStore.kt
package tech.mmarca.openvitals.devices.garmin

import java.io.File
import java.time.Instant
import kotlin.time.Duration
import kotlin.time.Duration.Companion.days

/**
 * Keeps the raw FIT files pulled off a watch. Archiving is destructive:
 * the watch never offers an archived file again, so bytes land on disk
 * before the archive flag is sent. Also lets a parser bug be reproduced.
 */
class GarminFileStore(
    /** Resolves the directory on first use. A callback, so a test can point it at a temp dir. */
    private val resolveDirectory: suspend () -> File,
    /** How long a file is kept before [prune] removes it. */
    private val retention: Duration = 30.days,
) {

    /** The note beside each saved file that is not imported yet, by the file object it belongs to. */
    private val pendingNotes = java.util.Collections.synchronizedMap(
        java.util.IdentityHashMap<GarminDownloadedFile, File>(),
    )

    /**
     * Writes [file] and returns its path. Throws, so the caller can refuse to archive.
     *
     * Also leaves a note that the file awaits import for [deviceId]. The watch archives a
     * file once it is saved and never offers it again, and the import runs only after the
     * whole pull. A crash or an exception in between used to lose the file: nothing read
     * the store back. [pending] does, and [markImported] removes the note.
     */
    suspend fun save(file: GarminDownloadedFile, now: Instant, deviceId: String): String {
        val directory = resolveDirectory()
        directory.mkdirs()
        // The timestamp keeps a re-download from clobbering an earlier copy: several files share number 65535.
        val name = "${file.entry.type.label}_${file.entry.fileIndex}_" +
            "${now.toEpochMilli()}.fit"
        val target = File(directory, name)
        target.writeBytes(file.bytes)
        val note = File(directory, name + PENDING_SUFFIX)
        note.writeText(file.entry.toPendingNote(deviceId))
        pendingNotes[file] = note
        GarminLog.log("[GARMIN-STORE] saved $name (${file.bytes.size}B)")
        return target.path
    }

    /** Files saved for [deviceId] that no import has finished with, oldest first. */
    suspend fun pending(deviceId: String): List<GarminDownloadedFile> {
        val directory = resolveDirectory()
        val notes = directory.listFiles().orEmpty()
            .filter { it.isFile && it.name.endsWith(PENDING_SUFFIX) }
            .sortedBy { it.lastModified() }
        return notes.mapNotNull { note ->
            val fields = runCatching { note.readLines() }.getOrNull()
                ?.mapNotNull { line -> line.split('=', limit = 2).takeIf { it.size == 2 } }
                ?.associate { it[0] to it[1] }
            if (fields == null || fields["device"] != deviceId) return@mapNotNull null
            val data = File(directory, note.name.removeSuffix(PENDING_SUFFIX))
            val entry = fields.toDirectoryEntry()
            if (entry == null || !data.isFile) {
                // Nothing to import from: the bytes are gone or the note is unreadable.
                note.delete()
                return@mapNotNull null
            }
            GarminDownloadedFile(entry, data.readBytes()).also { pendingNotes[it] = note }
        }
    }

    /** The import is done with [files], whatever it made of them. Their notes go. */
    fun markImported(files: List<GarminDownloadedFile>) {
        files.forEach { file -> pendingNotes.remove(file)?.delete() }
    }

    /** Deletes files older than [retention]. Best-effort. */
    suspend fun prune(now: Instant) {
        val cutoff = now.minusMillis(retention.inWholeMilliseconds)
        try {
            val directory = resolveDirectory()
            if (!directory.exists()) return
            for (entity in directory.listFiles().orEmpty()) {
                if (!entity.isFile) continue
                if (!entity.name.endsWith(".fit") && !entity.name.endsWith(PENDING_SUFFIX)) continue
                if (Instant.ofEpochMilli(entity.lastModified()).isBefore(cutoff)) {
                    entity.delete()
                    GarminLog.log("[GARMIN-STORE] pruned ${entity.path}")
                }
            }
        } catch (error: Exception) {
            GarminLog.log("[GARMIN-STORE] prune failed: $error")
        }
    }

    private companion object {
        const val PENDING_SUFFIX = ".pending"
    }
}

private fun GarminDirectoryEntry.toPendingNote(deviceId: String): String = listOf(
    "device" to deviceId,
    "dataType" to type.dataType,
    "subType" to type.subType,
    "fileIndex" to fileIndex,
    "fileNumber" to fileNumber,
    "specificFlags" to specificFlags,
    "fileFlags" to fileFlags,
    "fileSize" to fileSize,
    "fileDate" to (fileDate?.epochSecond ?: ""),
    "remoteDedupKey" to (remoteDedupKey ?: ""),
).joinToString("\n") { (key, value) -> "$key=$value" }

private fun Map<String, String>.toDirectoryEntry(): GarminDirectoryEntry? {
    val type = GarminFileType.fromCodes(
        get("dataType")?.toIntOrNull() ?: return null,
        get("subType")?.toIntOrNull() ?: return null,
    ) ?: return null
    return GarminDirectoryEntry(
        fileIndex = get("fileIndex")?.toIntOrNull() ?: return null,
        type = type,
        fileNumber = get("fileNumber")?.toIntOrNull() ?: return null,
        specificFlags = get("specificFlags")?.toIntOrNull() ?: 0,
        fileFlags = get("fileFlags")?.toIntOrNull() ?: 0,
        fileSize = get("fileSize")?.toLongOrNull() ?: return null,
        fileDate = get("fileDate")?.toLongOrNull()?.let(Instant::ofEpochSecond),
        remoteDedupKey = get("remoteDedupKey")?.takeIf { it.isNotEmpty() },
    )
}
